### PR TITLE
Encodes the last election certificate IDs in the batch header and subdag

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

This PR encodes the last election certificate IDs in the batch header and subdag, with backwards compatibility.

The sister PR for this is is here: https://github.com/AleoHQ/snarkOS/pull/2891

### CODE REVIEW

**Review by ljedrz — APPROVED** [review]

File: ledger/narwhal/batch-header/src/bytes.rs [hunk 1]
```diff
@@ -20,7 +20,7 @@ impl<N: Network> FromBytes for BatchHeader<N> {
         // Read the version.
         let version = u8::read_le(&mut reader)?;
         // Ensure the version is valid.
-        if version != 1 {
+        if version != 1 && version != 2 {
```
Comment (niklaslong) [inline ledger/narwhal/batch-header/src/bytes.rs:24]: Should these be constants?

**Review by niklaslong — APPROVED** [review]

File: ledger/narwhal/batch-header/src/bytes.rs [hunk 2]
```diff
@@ -65,12 +65,44 @@ impl<N: Network> FromBytes for BatchHeader<N> {
             previous_certificate_ids.insert(Field::read_le(&mut reader)?);
         }
 
+        // TODO (howardwu): For mainnet - Change this to always encode the number of committed certificate IDs.
+        //  We currently only encode the size and certificates only in the new version, for backwards compatibility.
```
Comment (acoglio) [inline ledger/narwhal/batch-header/src/bytes.rs]:
```suggestion
        //  We currently only encode the size and certificates in the new version, for backwards compatibility.
```

File: ledger/narwhal/batch-header/src/bytes.rs [hunk 3]
```diff
@@ -107,6 +140,19 @@ impl<N: Network> ToBytes for BatchHeader<N> {
             // Write the certificate ID.
             certificate_id.write_le(&mut writer)?;
         }
+        // TODO (howardwu): For mainnet - Change this to always encode the number of committed certificate IDs.
+        //  We currently only encode the size and certificates only in the new version, for backwards compatibility.
```
Comment (acoglio) [inline ledger/narwhal/batch-header/src/bytes.rs]:
```suggestion
        //  We currently only encode the size and certificates in the new version, for backwards compatibility.
```

**Review by acoglio — APPROVED** [review]

File: ledger/narwhal/batch-header/src/bytes.rs (hunk 1 above)
Comment (howardwu) [reply ledger/narwhal/batch-header/src/bytes.rs:24]: I forgot to put a `// TODO: For mainnet` comment on this. This is only a temporary change.

Comment (howardwu) [reply ledger/narwhal/batch-header/src/bytes.rs:24]: Fixed in dfd500b5a

**Pull request merged**